iroh-gateway/src/core.rs

```diff
@@ -0,0 +1,412 @@
+    constants::*,
+    error::GatewayError,
+    response::{GatewayResponse, ResponseFormat},
+};
+
+pub struct Core {
+    pub config: Config,
+    client: Client,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct GetParams {
+    format: Option<String>,
+    filename: Option<String>,
+    download: Option<String>,
+}
+
+impl Core {
+    pub fn new(config: Config) -> Self {
+        Self {
+            config,
+            client: Client::new(),
+        }
+    }
+
+    pub async fn serve(&self) {
+        let app = Router::new()
+            .route("/ipfs/:cid", get(get_ipfs))
+            .route("/ipfs/:cid/*cpath", get(get_ipfs))
+            .layer(Extension(self.config.clone()))
+            .layer(Extension(self.client.clone()))
+            .layer(
+                ServiceBuilder::new()
+                    // Handle errors from middleware
+                    .layer(HandleErrorLayer::new(middleware_error_handler))
+                    .load_shed()
+                    .concurrency_limit(1024)
+                    .timeout(Duration::from_secs(10))
+                    .into_inner(),
+            );
+        let addr = format!("0.0.0.0:{}", self.config.port);
+        axum::Server::bind(&addr.parse().unwrap())
+            .http1_preserve_header_case(true)
+            .http1_title_case_headers(true)
+            .serve(app.into_make_service())
+            .await
+            .unwrap();
+    }
+}
+
+async fn get_ipfs(
+    Extension(config): Extension<Config>,
+    Extension(client): Extension<Client>,
+    Path(params): Path<HashMap<String, String>>,
+    Query(query_params): Query<GetParams>,
+) -> Result<GatewayResponse, GatewayError> {
+    // parse path params
+    let cid = params.get("cid").unwrap();
+    let cpath = "".to_string();
+    let cpath = params.get("cpath").unwrap_or(&cpath);
+    let full_content_path = format!("/ipfs/{}{}", cid, cpath);
+
+    // parse query params
+    let format = query_params.format.unwrap_or("".to_string());
+    let format = ResponseFormat::try_from(format.as_str());
+    let format = match format {
+        Ok(f) => f,
+        Err(err) => {
+            let msg = format!("{}", err);
+            return error(StatusCode::BAD_REQUEST, &msg);
+        }
+    };
+    let query_file_name = query_params.filename.unwrap_or("".to_string());
+    let download = query_params.download.unwrap_or("".to_string()) == "true";
+
+    // init headers
+    let mut headers = HashMap::new();
+    format.write_headers(&mut headers);
+    headers.insert(HEADER_X_IPFS_PATH.to_string(), full_content_path.clone());
+    add_user_headers(&mut headers, config.headers.clone());
+
+    // handle request and fetch data
+    let req = Request {
+        format,
+        cid: cid.to_string(),
+        full_content_path,
+        query_file_name,
+        content_path: cpath.to_string(),
+        download,
+    };
+    match req.format {
+        ResponseFormat::Raw => serve_raw(&req, client, &mut headers).await,
```

Comment:
the headers are not used again so `serve_raw` should take ownership of them

Reply:
Unsure what you mean here, headers are used/changed in those calls (even in this simple stub form)

Reply:
But they are not returned/used in this call anymore, asfaict.